src/coins.cpp

```diff
@@ -260,6 +299,265 @@ void CCoinsViewCache::BringBestAnchorIntoCache(
     assert(GetSaplingAnchorAt(currentRoot, tree));
 }
 
+void draftMMRNode(std::vector<uint32_t> &indices,
+                  std::vector<HistoryEntry> &entries,
+                  HistoryNode nodeData,
+                  uint32_t alt,
+                  uint32_t peak_pos)
+{
+    HistoryEntry newEntry = alt == 0
+        ? libzcash::LeafToEntry(nodeData)
+        // peak_pos - (1 << alt) is the mmr position of left child, -1 to that is this position of entry in
+        // array representation.
+        //
+        // peak_pos - 1 is the mmr position of right child, -1 to that is this position of entry in
+        // array representation
+        : libzcash::NodeToEntry(nodeData, peak_pos - (1 << alt) - 1, peak_pos - 2);
+
+    indices.push_back(peak_pos - 1);
+    entries.push_back(newEntry);
+}
+
+// Computes floor(log2(x)).
+static inline uint32_t floor_log2(uint32_t x) {
+    assert(x > 0);
+    int log = 0;
+    while (x >>= 1) { ++log; }
+    return log;
+}
+
+// Computes the altitude of the largest subtree for an MMR with n nodes,
+// which is floor(log2(n + 1)) - 1.
+static inline uint32_t altitude(uint32_t n) {
+    return floor_log2(n + 1) - 1;
+}
+
+uint32_t CCoinsViewCache::PreloadHistoryTree(uint32_t epochId, bool extra, std::vector<HistoryEntry> &entries, std::vector<uint32_t> &entry_indices) {
+    auto treeLength = GetHistoryLength(epochId);
+
+    if (treeLength <= 0) {
+        throw std::runtime_error("Invalid PreloadHistoryTree state called - tree should exist");
+    } else if (treeLength == 1) {
+        entries.push_back(libzcash::LeafToEntry(GetHistoryAt(epochId, 0)));
+        entry_indices.push_back(0);
+        return 1;
+    }
+
+    uint32_t last_peak_pos = 0;
+    uint32_t last_peak_alt = 0;
+    uint32_t alt = 0;
+    uint32_t peak_pos = 0;
+    uint32_t total_peaks = 0;
+
+    // Assume the following example peak layout with 14 leaves, and 25 nodes in total:
+    //
+    //             P
+    //            /\
+    //           /  \
+    //          / \  \
+    //        /    \  \  Altitude
+    //     _A_      \  \    3
+    //   _/   \_     B  \   2
+    //  / \   / \   / \  C  1
+    // /\ /\ /\ /\ /\ /\ /\ 0
+    //
+    // We start by determining the altitude of the highest peak (A).
+    alt = altitude(treeLength);
+
+    // We determing the position of the highest peak (A) by pretending it is the right
+    // sibling in a tree, and its left-most leaf has position 1. Then the left sibling
+    // of (A) has position 0, and so we can "jump" to the peak's position by computing
+    // 0 + 2^(alt + 1) - 1. This is also why peak_pos is 1-indexed here; we subtract 1
+    // when we index into the underlying array structure.
+    peak_pos = (1 << (alt + 1)) - 1;
```

Comment:
Alternatively we could subtract 1 here so that after we've used this trick to find the highest peak's position, the rest of the code doesn't need to account for 1-based-indexing. I was confused when I read `draftMMRNode()` before reading this code.